tests/bdd.py

```diff
@@ -1,83 +1,129 @@
 
-from event_sourcery.event_store import Event, EventStore, Metadata, StreamId
-from tests.factories import next_version
+from event_sourcery import event_store as es
+from event_sourcery.event_store import Event, Position, Recorded
 from tests.matchers import any_metadata
 
 
 @dataclass
 class Stream:
-    store: EventStore
-    id: StreamId = field(default_factory=StreamId)
+    store: es.EventStore
+    id: es.StreamId = field(default_factory=es.StreamId)
 
-    def receives(self, *events: Metadata) -> Self:
+    @singledispatchmethod
+    def receives(self, *events: es.Metadata) -> Self:
+        self.autoversion(*events)
         self.store.append(*events, stream_id=self.id)
         return self
 
+    @receives.register
+    def receives_base_events(self, *events: es.Event) -> Self:
+        metadata = (es.Metadata.wrap(e, version=None) for e in events)
+        return self.receives(*metadata)
+
+    def snapshots(self, snapshot: es.Metadata) -> Self:
+        if not snapshot.version:
+            snapshot.version = self.current_version
+        self.store.save_snapshot(self.id, snapshot)
+        return self
+
     @property
-    def events(self) -> list[Metadata]:
+    def events(self) -> list[es.Metadata]:
         return list(self.store.load_stream(self.id))
 
-    def loads_only(self, events: Sequence[Metadata]) -> None:
+    def loads_only(self, events: Sequence[es.Metadata]) -> None:
         assert self.events == list(events)
 
-    def loads(self, events: Sequence[Metadata] | Sequence[Event]) -> None:
-        if not all([isinstance(e, Metadata) for e in events]):
-            events = [any_metadata(e) for e in cast(Sequence[Event], events)]
+    def loads(self, events: Sequence[es.Metadata | es.Event]) -> None:
+        events = [e if isinstance(e, es.Metadata) else any_metadata(e) for e in events]
         assert self.events == list(events)
 
     def is_empty(self) -> None:
         assert self.events == []
 
+    @property
+    def current_version(self) -> int | None:
+        return (self.events or [Mock(version=0)])[-1].version
+
+    def autoversion(self, *events: es.Metadata) -> None:
+        if any(e.version is not None for e in events):
+            return
+
+        if (current_version := self.current_version) is None:
+            return
+
+        for version, e in enumerate(events, start=current_version + 1):
+            e.version = version
+
 
 @dataclass
-class Given:
-    store: EventStore
+class Subscription:
+    _subscription: Iterator[Recorded] | Iterator[es.Entry]
+
+    def next_received_record_is(self, expected: Recorded | es.Entry) -> None:
+        received = next(self._subscription)
+        assert expected == received
+
 
-    def stream(self, with_id: StreamId | None = None) -> Stream:
+T = TypeVar("T")
+
+
+@dataclass
+class Step:
+    store: es.EventStore
+
+    def __call__(self, value: T) -> T:
+        return value
+
+    def subscription(
+        self,
+        to: Position | None = None,
```

Comment:
Why `to` is passed to `from_position`? that's counterintuitive

Reply:
I think when I was writing tests I wanted a shorter version in DSL.

Reply:
My point is that `to` is completely opposite in terms of meaning to `from` and that confused me